### PR TITLE
PDCT-1115 Change from the repository pattern to active objects (like sqlalchemy)

### PR DESCRIPTION
**Pull request by diversemix**
# Description

- Renames with_transaction to with_database, removing the context which is now private to the decorator.

## TODO

- bump version
- remove the transaction for the decorator

## Proposed version

Please select the option below that is most relevant from the list below. This
will be used to generate the next tag version name during auto-tagging.

- [ ] Skip auto-tagging
- [x] Patch
- [ ] Minor version
- [ ] Major version

Visit the [Semver website](https://semver.org/#summary) to understand the
difference between `MAJOR`, `MINOR`, and `PATCH` versions.

Notes:

- If none of these options are selected, auto-tagging will fail
- Where multiple options are selected, the most senior option ticked will be
  used -- e.g. Major > Minor > Patch
- If you are selecting the version in the list above using the textbox, make
  sure your selected option is marked `[x]` with no spaces in between the
  brackets and the `x`

## Type of change

Please select the option(s) below that are most relevant:

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] GitHub workflow update
- [ ] Documentation update
- [ ] Remove legacy code
- [ ] Dependency update

## How Has This Been Tested?

Please describe the tests that you added to verify your changes.

## Reviewer Checklist

- [ ] The PR represents a single feature (small drive-by fixes are also ok)
- [ ] The PR includes tests that are sufficient for the level of risk
- [ ] The code is sufficiently commented, particularly in hard-to-understand areas
- [ ] Any required documentation updates have been made
- [ ] Any TODOs added are captured in future tickets
- [ ] No FIXMEs remain

### CODE REVIEW

File: app/service/collection.py [hunk 1]
```diff
@@ -136,22 +137,32 @@ def update(
...
+        db = db_session.get_db()
+
     # TODO: implement changing of a collection's organisation
     # org_id = organisation.get_id_from_name(db, collection.organisation)
 
     validate_import_id(import_id)
 
-    collection_repo.update(db, import_id, collection)
-    db.commit()
+    transaction = db.begin_nested()
```
Comment (diversemix) [inline app/service/collection.py:148]: @katybaulch  this will be the pattern we are to expect in delete/update functions - are you ok with this?

File: app/repository/event.py [hunk 2]
```diff
@@ -166,28 +166,23 @@ def create(db: Session, event: EventCreateDTO) -> str:
     :return str: The import id of the newly created family event.
     """
 
-    try:
-        new_family_event = _event_from_dto(event)
-
-        family_import_id = new_family_event.family_import_id
+    new_family_event = _event_from_dto(event)
```
Comment (diversemix) [inline app/repository/event.py:169]: Eeek - I remember doing this ... the handler needs putting back!

File: app/service/collection.py [hunk 3]
```diff
@@ -109,24 +107,26 @@ def validate_multiple_ids(
 
 
 @validate_call(config=ConfigDict(arbitrary_types_allowed=True))
-def validate(import_ids: set[str], db: Session = db_session.get_db()) -> None:
+def validate(import_ids: set[str], db: Optional[Session]) -> None:
```
Comment (diversemix) [inline app/service/collection.py:110]: FYI: The function signatures for all services were changed so that `get_db` was not called when the module was loaded ... this is deferred to inside the function if `None`

File: app/service/event.py [hunk 4]
```diff
@@ -96,25 +94,34 @@ def create(
...
-
-
-@db_session.with_transaction(__name__)
+    try:
+        import_id = event_repo.create(db, event)
+        if len(import_id) == 0:
+            db.rollback()
+        return import_id
+    except Exception as e:
+        # db.rollback()
```
Comment (diversemix) [inline app/service/event.py:113]: I need to see if I can put this back

File: tests/integration_tests/conftest.py [hunk 5]
```diff
@@ -107,26 +107,32 @@ def data_db_connection() -> Generator[Connection, None, None]:
...
-        return session
-
-    monkeypatch.setattr(db_session, "get_db", get_test_db)
-
-    yield session
-
-    session.close()
-    print(f"This test is finished and being rolledback with transaction {transaction}")
-    transaction.rollback()
+def data_db(slow_db):
```
Comment (diversemix) [inline tests/integration_tests/conftest.py:110]: Old function commented out below - could add a comment - but I think we both know the history of this ... pain :crossed_swords:

File: tests/integration_tests/document/test_create.py [hunk 6]
```diff
@@ -362,7 +362,7 @@ def test_create_document_when_invalid_variant(
     )
     assert response.status_code == status.HTTP_503_SERVICE_UNAVAILABLE
     data = response.json()
-    assert '(Invalid) is not present in table "variant"' in data["detail"]
+    assert len(data["detail"]) > 0
```
Comment (diversemix) [inline tests/integration_tests/document/test_create.py:365]: We have lost some of the readability of errors with the new way of doing context

File: app/repository/document.py [hunk 7]
```diff
@@ -132,7 +132,7 @@ def _doc_to_dto(doc_query_return: ReadObj) -> DocumentReadDTO:
         type=str(fdoc.document_type) if fdoc.document_type is not None else None,
         created=cast(datetime, fdoc.created),
         last_modified=cast(datetime, fdoc.last_modified),
-        slug=str(fdoc.slugs[-1].name if len(fdoc.slugs) > 0 else ""),
+        slug=str(fdoc.slugs[0].name if len(fdoc.slugs) > 0 else ""),
```
Comment (katybaulch) [inline app/repository/document.py:135]: Nice driveby!

File: app/repository/document.py [hunk 8]
```diff
@@ -404,6 +404,7 @@ def create(db: Session, document: DocumentCreateDTO) -> str:
 
         org_name = cast(str, org.name)
 
+        # TODO: Fix this call - the transaction is changed here
```
Comment (katybaulch) [inline app/repository/document.py:407]: Do we still need this? If so, can we link a ticket please?

File: app/service/collection.py (hunk 3 above)
Comment (katybaulch) [inline app/service/collection.py:110]: This is going to make Bandit so so so happy - the db call in the function signature was a vulnerability!

File: app/service/collection.py [hunk 9]
```diff
@@ -166,15 +172,25 @@ def create(
...
+        db = db_session.get_db()
 
-@db_session.with_transaction(__name__)
+    try:
+        import_id = collection_repo.create(db, collection, user.org_id)
+        if len(import_id) == 0:
+            db.rollback()
+        return import_id
+    except Exception as e:
+        # db.rollback()
```
Comment (katybaulch) [inline app/service/collection.py:185]: Commented out code?

File: tests/integration_tests/conftest.py (hunk 5 above)
Comment (katybaulch) [reply tests/integration_tests/conftest.py:110]: Do we want to keep it commented out?

**Review by katybaulch — CHANGES_REQUESTED** [review]
Looking insaneeeee - nice drivebys on the slugs as well ✨  Couple of bits to tidy, but all good otherwise :)

File: app/repository/document.py (hunk 7 above)
Comment (diversemix) [reply app/repository/document.py:135]: ta :bow:

File: app/service/collection.py (hunk 9 above)
Comment (diversemix) [reply app/service/collection.py:185]: hmmm -- thought that was restored - let me fix

File: app/repository/document.py (hunk 8 above)
Comment (diversemix) [reply app/repository/document.py:407]: yeah that would be good ...

Comment (diversemix) [reply app/repository/document.py:407]: ... okay I've removed the comment. But if we do want to start using the quick tests again - then we need to ensure that we pass the current transaction into the `db-client` ... also creating a ticket...

Comment (diversemix) [reply app/repository/document.py:407]: https://linear.app/climate-policy-radar/issue/PDCT-1197/possible-rollback-issue

File: tests/integration_tests/conftest.py (hunk 5 above)
Comment (diversemix) [reply tests/integration_tests/conftest.py:110]: I think so for now - it keeps out thoughts alive - I still would like to do more work on this when I can... but if you insist I;'ll remove

**Review by katybaulch — APPROVED** [review]
happy with Peter's approach to come back and tidy this later, as we're blocked on merging our other PRs at the moment
